temba/contacts/models.py

```diff
@@ -1020,11 +1022,79 @@ def set_field(self, user, key, value, label=None, importing=False):
                         )
                     )
 
-        # update any groups or campaigns for this contact if not importing
-        if has_changed and not importing:
-            self.handle_update(field=field)
+        # update any groups or campaigns for this contact
+        if has_changed:
+            self.handle_update(fields=[field.key])
+
+    def set_fields(self, user, fields):
+        if self.fields is None:
+            self.fields = {}
+
+        fields_for_delete = set()
+        fields_for_update = set()
+        all_fields = {}
+        changed_field_keys = set()
 
-    def handle_update(self, attrs=(), urns=(), field=None, group=None, is_new=False):
+        for key, value in fields.items():
+            field = ContactField.get_or_create(self.org, user, key)
+
+            field_uuid = str(field.uuid)
+
+            # parse into the appropriate value types
+            if value is None or value == "":
+                # setting a blank value is equivalent to removing the value
+                value = None
+
+                # value being cleared, remove our key
+                if field_uuid in self.fields:
+                    fields_for_delete.add(field_uuid)
+
+                    changed_field_keys.add(key)
+
+            else:
+                field_dict = self.serialize_field(field, value)
+
+                # update our field if it is different
+                if self.fields.get(field_uuid) != field_dict:
+                    fields_for_update.add(field_uuid)
+                    all_fields.update({field_uuid: field_dict})
+
+                    changed_field_keys.add(key)
+
+        modified_by_id = user.id
+        modified_on = timezone.now()
+
+        # if there was a change, update our JSONB on our contact
+        if fields_for_delete:
+            with connection.cursor() as cursor:
+                # prepare expression for multiple field delete
+                remove_fields = " - ".join(f"'{field_uuid}'" for field_uuid in fields_for_delete)
+                cursor.execute(
+                    f"UPDATE contacts_contact SET fields = fields - {remove_fields}, modified_by_id = %s, modified_on = %s WHERE id = %s",
+                    [modified_by_id, modified_on, self.id],
+                )
+
+        if fields_for_update:
+            with connection.cursor() as cursor:
+                cursor.execute(
+                    "UPDATE contacts_contact SET fields = COALESCE(fields,'{}'::jsonb) || %s::jsonb, modified_by_id = %s, modified_on = %s WHERE id = %s",
+                    [json.dumps(all_fields), modified_by_id, modified_on, self.id],
+                )
+
+        # update local contact cache
```

Comment:
this is a bit strange, not sure if it's only required for tests, but we do need to update the internal object cache which is used later in the code ... I don't think that this is ideal